R/drive_list.R

```diff
@@ -0,0 +1,191 @@
+#'   I wanted 200, I would run `drive_ls(pageSize = 200)`.
+
+#' Helpful links for forming queries:
+#'   * <https://developers.google.com/drive/v3/web/search-parameters>
+#'   * <https://developers.google.com/drive/v3/reference/files/list>
+
+#' @return tibble containing the name, type, and id of files on your google
+#'   drive (default 100 files)
+#' @examples
+#' \dontrun{
+#' ## list user's Google Sheets
+#' drive_list(q = "mimeType='application/vnd.google-apps.spreadsheet'")
+#' }
+#'
+#' @export
+drive_list <- function(path = NULL, pattern = NULL, ..., verbose = TRUE){
+
+  if (!is.null(pattern)) {
+    if (!(inherits(pattern, "character") && length(pattern) == 1)) {
+      stop("Please update `pattern` to be a character string.", call. = FALSE)
+    }
+  }
+
+  x <- list(...)
+
+  if (is.null(x$fields)) {
+    x$fields <- paste0("files/", .drive$default_fields, collapse = ",")
+  }
+
+  if (!is.null(path)) {
+    folder <- get_leafmost_id(path = path)
+    q <- paste0("'", folder, "'", " in parents")
+    if (is.null(x$q)) {
+      x$q <- q
+    } else {
+      x$q <- paste(x$q, "and", q)
+    }
+  }
+
+  request <- build_request(
+    endpoint = .drive$base_url_files_v3,
+    token = drive_token(),
+    params = x
+  )
+  response <- make_request(request)
+  proc_res <- process_request(response)
+
+  req_tbl <- tibble::tibble(
+    name = purrr::map_chr(proc_res$files, "name"),
+    type = sub(".*\\.", "", purrr::map_chr(proc_res$files, "mimeType")),
+    ## FIXME: all parents should be taken, not just the first
+    ## parents should be a list-column
+    parents = purrr::map_chr(proc_res$files, list("parents", 1), .null = NA),
+    id = purrr::map_chr(proc_res$files, "id"),
+    gfile = proc_res$files
+  )
+
+  if (is.null(pattern)) {
+    return(req_tbl)
+  }
+
+  keep_names <- grep(pattern, req_tbl$name)
+  if (length(keep_names) == 0L) {
+    if (verbose) message(sprintf("No file names match the pattern: '%s'.", pattern))
+    return(invisible())
+  }
+  req_tbl[keep_names, ]
+}
+
+get_leafmost_id <- function(path) {
+
+  path_pieces <- unlist(strsplit(path, "/"))
+  d <- length(path_pieces)
+  path_pattern <- paste0("^", path_pieces, "$", collapse = "|")
+  folders <- drive_list(
+    pattern = path_pattern,
+    fields = "files/parents,files/name,files/mimeType,files/id",
+    q = "mimeType='application/vnd.google-apps.folder'"
+  )
+  ## FIXME
+  ## seems like the "parents" variable should be a list-column?
+  ## can't parents technically have length greater than one?
+
+  if (!all(path_pieces %in% folders$name)){
+    spf("We could not find the file path '%s' on your Google Drive", path)
+  }
+  ## guarantee: we have found at least one folder with correct name for
+  ## each piece of path
+
+  folders$depth <- match(folders$name, path_pieces)
+  folders <- folders[order(folders$depth), ]
+  folder <- folders$id[folders$depth == d]
+  if (length(folder) != 1) {
```

Comment:
I think this would eliminate the case where you may have `foo/bar/baz` and also `foo/yo/baz`, which should be allowed, right?